src/backend/commands/copyfromparse.c

```diff
@@ -1735,3 +1785,92 @@ CopyReadBinaryAttribute(CopyFromState cstate, FmgrInfo *flinfo,
 	*isnull = false;
 	return result;
 }
+
+void
+RemoveInvalidDataInBuf(CopyFromState cstate)
+{
+	int nbytes;
+	int scanidx;
+
+	if (cstate->errMode == ALL_OR_NOTHING)
+		CopyConversionError(cstate);
+
+	/* reach EOF, do not need clean */
+	if (cstate->input_reached_eof)
+		return;
+
+	if (!cstate->need_transcoding)
+	{
+		/*
+		 * According to `BeginCopyFrom`, if not need_transcoding these two
+		 * pointer share one memory space.
+		 */
+		Assert(cstate->raw_buf == cstate->input_buf);
+
+		if (FindEolInUnverifyRawBuf(cstate, &scanidx))
+		{
+			cstate->raw_buf_index += scanidx;
+			nbytes = RAW_BUF_BYTES(cstate);
+			if (nbytes > 0 && cstate->raw_buf_index > 0)
+				memmove(cstate->raw_buf,
+						cstate->raw_buf + cstate->raw_buf_index, nbytes);
+
+			cstate->raw_buf_len -= cstate->raw_buf_index;
+			cstate->raw_buf_index = 0;
+		}
+		else
+		{
+			/* Current page can not find eol, to skip current raw buffer */
+			cstate->raw_buf_len = 0;
+			cstate->raw_buf_index = 0;
+
+			/* leave a hint to identify find eol after next raw page read */
+			cstate->find_eol_with_rawreading = true;
+		}
+
+		/* reset input buf, so we can redo conversion/verification */
+		cstate->input_reached_error = false;
+		cstate->input_buf_index = 0;
+		cstate->input_buf_len = 0;
+
+		/* reset line_buf */
+		resetStringInfo(&cstate->line_buf);
+		cstate->line_buf_valid = false;
+		cstate->cdbsreh->rejectcount++;
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("Data validation error: since the source data "
+							   "need transcoding sreh can not handle yet.")));
+	}
+}
+
+static bool
+FindEolInUnverifyRawBuf(const CopyFromState cstate, int *scanidx)
+{
+	EolType eol_type = cstate->opts.eol_type;
+	bool found = false;
+	for (*scanidx = 0;
+		 *scanidx < (cstate->raw_buf_len - cstate->raw_buf_index);
+		 (*scanidx)++)
+	{
+		char c1 = cstate->raw_buf[cstate->raw_buf_index + *scanidx];
+		char c2 = cstate->raw_buf[cstate->raw_buf_index + *scanidx + 1];
+		if (((eol_type == EOL_CRNL || eol_type == EOL_UNKNOWN) && c1 == '\r' &&
+			 c2 == '\n'))
```

Comment:
Is it possible index `*scanidx + 1` beyond the bound of raw_buf?

Reply:
nice cache!